libraries/lib-wave-track/SequenceSampleCache.cpp

```diff
@@ -0,0 +1,77 @@
+#include "SequenceSampleCache.h"
+
+#include <cassert>
+
+SequenceSampleCache::SequenceSampleCache(
+   SourceSampleGetterFn fn, sampleCount sequenceLength, size_t maxBlockSize)
+    : mGetterFn(std::move(fn))
+    , mMaxBlockSize(maxBlockSize)
+    , mSequenceLength(sequenceLength)
+{
+}
+
+bool SequenceSampleCache::Get(
+   samplePtr buffer, sampleFormat format, sampleCount start, size_t len,
+   bool mayThrow)
+{
+   auto containingBlockIt = std::find_if(
+      mBlocks.begin(), mBlocks.end(),
+      [start](const std::optional<Block>& block) {
+         return block.has_value() && block->Contains(start);
+      });
+   if (containingBlockIt == mBlocks.end())
+   {
+      const auto blockIndex =
+         mLastContainingBlockIndex.has_value() ?
+            (*mLastContainingBlockIndex + 1) % mBlocks.size() :
+            0u;
+      const auto numSamplesToCache =
+         std::min(sampleCount { mMaxBlockSize }, mSequenceLength - start)
+            .as_size_t();
+      auto& block = mBlocks[blockIndex];
+      block.emplace(numSamplesToCache, format, start);
+      if (!mGetterFn(
+             block->buffer.ptr(), format, start, numSamplesToCache, mayThrow))
+      {
+         assert(false);
+         std::fill(
+            block->buffer.ptr(),
+            block->buffer.ptr() + numSamplesToCache * SAMPLE_SIZE(format), 0);
+      }
+      mLastContainingBlockIndex = blockIndex;
+      containingBlockIt = mBlocks.begin() + blockIndex;
+   }
+   const auto& block = *containingBlockIt;
+   const auto sampleOffset = (start - block->sequenceStartSample).as_size_t();
+   const auto numSamplesToRead =
+      std::min(len, block->numSamples - sampleOffset);
+   const auto numBytesToRead = numSamplesToRead * SAMPLE_SIZE(format);
+   const auto blockPtr =
+      block->buffer.ptr() + sampleOffset * SAMPLE_SIZE(format);
+   std::copy(blockPtr, blockPtr + numBytesToRead, buffer);
+   if (numSamplesToRead == len)
+   {
+      return true;
+   }
+   else
+   {
+      return Get(
+         buffer + numBytesToRead, format, start + numSamplesToRead,
+         len - numSamplesToRead, mayThrow);
+   }
+}
```

Comment:
I have now read all of this replacement for `SampleTrackCache`.

That class had many more complications that I wrote long ago and reviewed again just today.  Did you read and understand all of it?  I don't ask to reproduce all of those, yet..

Like this cache, that one was meant to work best with a pattern of fetching from successive sample offsets.  Many small fetches then would not be punished with repeated database queries.

Unlike this, that one also allowed for efficient, sequentially backward fetches, as might also happen in scrubbing.

Also unlike this, that one had awareness of the underlying block boundaries within the Sequence and adjusted its caching pre-fetch sizes accordingly.  But that is lost with the simplifications here.

Removing the complications of best and ideal block sizes from WaveTrack and WaveClip interfaces still seems a good idea to me, cleaning up those abstractions, but now that the new cache can cooperate more intimately with Sequence, it should still implement similar logic to align its prefetches to the underlying Sequence blocks.

Also unlike this, SampleTrackCache maintained two buffers and could satisfy a request partly from each.  So suppose two buffers equal to the maximum block size.

So suppose only the maximum block size is exposed by WaveTrack as a preferred size for sequential fetches, and suppose an Effect does just that.  (Let's neglect time stretching).  Still the request might not align to the actual blocks, because of the history of cuts and pastes causing some under-sized blocks, or because some starting point was chosen without regard to alignment.

But maintaining two buffers would still reduce the number of database queries.  A request for the maximum block size might be filled partly by one slot, then partly by the other, but then also the remainder of the other slot is available to satisfy the next request without again invoking the getter function.

In other words -- without reproducing all complications of `SampleTrackCache`, still I would rewrite the above to loop, filling alternate slots, then copying to the out buffer, until satisfying the total request.